File: dsa-problems/Graphs/solutions/java/GRP-009-city-toll-dijkstra.java
import java.util.*;

class Solution {
    public long[] dijkstra(int n, List<List<int[]>> adj, int source) {
        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[source] = 0;

        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        pq.offer(new long[]{0, source});

        while (!pq.isEmpty()) {
            long[] curr = pq.poll();
            long d = curr[0];
            int node = (int)curr[1];

            if (d > dist[node]) continue;

            List<int[]> neighbors = new ArrayList<>(adj.get(node));
            neighbors.sort((a, b) -> {
                if (a[1] != b[1]) return Integer.compare(a[1], b[1]);
                return Integer.compare(a[0], b[0]);
            });

            for (int[] edge : neighbors) {
                int neighbor = edge[0];
                long weight = edge[1];
                long newDist = dist[node] + weight;

                if (newDist < dist[neighbor]) {
                    dist[neighbor] = newDist;
                    pq.offer(new long[]{newDist, neighbor});
                }
            }
        }

        return dist;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        List<List<int[]>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int w = sc.nextInt();
            adj.get(u).add(new int[]{v, w});
        }

        // Handle optional source input
        int source = 0;
        if (sc.hasNextInt()) {
            source = sc.nextInt();
        }

        Solution solution = new Solution();
        long[] result = solution.dijkstra(n, adj, source);

        for (int i = 0; i < result.length; i++) {
            long val = result[i] == Long.MAX_VALUE ? -1 : result[i];
            System.out.print(val);
            if (i < result.length - 1) System.out.print(" ");
        }
        System.out.println();
        sc.close();
    }
}

File: dsa-problems/Graphs/solutions/javascript/GRP-009-city-toll-dijkstra.js
class MinHeap {
  constructor() {
    this.heap = [];
  }
  
  push(val) {
    this.heap.push(val);
    this.bubbleUp(this.heap.length - 1);
  }
  
  pop() {
    if (this.heap.length === 0) return null;
    const min = this.heap[0];
    const end = this.heap.pop();
    if (this.heap.length > 0) {
      this.heap[0] = end;
      this.bubbleDown(0);
    }
    return min;
  }
  
  bubbleUp(idx) {
    while (idx > 0) {
      const parent = Math.floor((idx - 1) / 2);
      if (this.heap[parent][0] <= this.heap[idx][0]) break;
      [this.heap[parent], this.heap[idx]] = [this.heap[idx], this.heap[parent]];
      idx = parent;
    }
  }
  
  bubbleDown(idx) {
    while (true) {
      let smallest = idx;
      const left = 2 * idx + 1;
      const right = 2 * idx + 2;
      
      if (left < this.heap.length && this.heap[left][0] < this.heap[smallest][0]) {
        smallest = left;
      }
      if (right < this.heap.length && this.heap[right][0] < this.heap[smallest][0]) {
        smallest = right;
      }
      if (smallest === idx) break;
      
      [this.heap[idx], this.heap[smallest]] = [this.heap[smallest], this.heap[idx]];
      idx = smallest;
    }
  }
  
  isEmpty() {
    return this.heap.length === 0;
  }
}

class Solution {
  dijkstra(n, adj, source) {
    const dist = new Array(n).fill(Infinity);
    dist[source] = 0;
    const pq = new MinHeap();
    pq.push([0, source]);
    
    while (!pq.isEmpty()) {
      const [d, node] = pq.pop();
      
      if (d > dist[node]) continue;
      adj[node].sort((a, b) => (a[1] - b[1]) || (a[0] - b[0]));
      for (const [neighbor, weight] of adj[node]) {
        const newDist = dist[node] + weight;
        if (newDist < dist[neighbor]) {
          dist[neighbor] = newDist;
          pq.push([newDist, neighbor]);
        }
      }
    }
    
    return dist.map((d) => (d === Infinity ? -1 : d));
  }
}

const fs = require('fs');

const data = fs.readFileSync(0, 'utf8').trim().split(/\s+/);
if (data.length > 0) {
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const m = parseInt(data[idx++], 10);
  const adj = Array.from({ length: n }, () => []);
  for (let i = 0; i < m; i++) {
    const u = parseInt(data[idx++], 10);
    const v = parseInt(data[idx++], 10);
    const w = parseInt(data[idx++], 10);
    adj[u].push([v, w]);
  }
  const source = idx < data.length ? parseInt(data[idx++], 10) : 0;
  const sol = new Solution();
  const result = sol.dijkstra(n, adj, source);
  console.log(result.join(' '));
}
